---
title: "v3.4"
author: "Steven N. Hart, Ph.D"
date: '7/10/2020'
output:
  html_document:
    keep_md: true
---

<style type="text/css">
.main-container {
  max-width: 1800px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(width = 1800)
knitr::opts_chunk$set(fig.width=12, fig.height=8, echo=TRUE, warning=FALSE, message=FALSE)

library(tidyverse)
library(data.table)
library(caret)
library(h2o)
library(pROC)
library(PRROC)
library(ggpubr)
library(e1071)
library(knitr)

# Change to TRUE to generate new models
train=FALSE
v_fold_nest = 10
training_time_limit=3600
set.seed(999)
```

```{r get-aligngvgd-data}
all_annotations = read.csv('sources/v3.4a.out',header=TRUE, stringsAsFactors = FALSE, sep="\t")

#Get BayesDel
bd = read.csv('sources/BayesDel_nsfp33a_noAF.tsv',header=TRUE, sep="\t", skip = 132)
bd$ID=NULL

names(bd) =c("Chr", "Pos", "Ref", "Alt", "BayesDel")
all_annotations = merge(all_annotations, bd, all.x=TRUE)
rm(bd)

#Get AlignGVGD
agvgd = read.csv('sources/alignGvgd.tsv', header=TRUE, sep="\t", stringsAsFactors = FALSE)
names(agvgd) =c("Chr", "Pos", "Ref", "Alt", "AlignGVGDPrior")
all_annotations = merge(all_annotations, agvgd, all.x=TRUE)
rm(agvgd)


for (j in names(all_annotations)[6:ncol(all_annotations)]){
    if (typeof(all_annotations[,j]) == 'character'){
      # Get the maximum score
      all_annotations[,j] = suppressWarnings(sapply(strsplit(all_annotations[,j], "\\|"), function(x) (max(as.numeric(x),na.rm = TRUE))) %>% as.numeric())
      # Remove -Inf
      all_annotations[which(all_annotations[,j] =='-Inf'),j] = NA
    }
}
######################################################
# Add my own AAPos Annotations
b1=read.csv('sources/CAVA_BRCA1.txt', header = FALSE, sep='\t', stringsAsFactors = FALSE)
b2=read.csv('sources/CAVA_BRCA2.txt', header = FALSE, sep='\t', stringsAsFactors = FALSE)
b3 = rbind(b1,b2)
names(b3) = c('CSN','Gene','Variant','ProtChange','RefProt','AApos','AltProt')
b3 = b3[,c('Gene','Variant','AApos')]
b3$AApos = as.numeric(b3$AApos)
all_annotations = all_annotations %>%
  mutate(Variant = paste(Chr,Pos,Ref,Alt,sep='_'), AApos=NULL)
all_annotations = merge(all_annotations, b3, by="Variant", all.x=TRUE)
######################################################


all_annotations = all_annotations %>%
  filter(AApos>0)

keep = names(all_annotations)[-grep("Rankscore",names(all_annotations))]
all_annotations = all_annotations[,keep]

rm(keep, j)
#glimpse(all_annotations)
```

Plot all predictions
```{r eval = FALSE, echo = FALSE}
scores = read.csv('sources/base_thresholds',sep="\t", header = FALSE)
names(scores) = 'Model'
p = NULL

for (model in scores$Model){
  tmp = all_annotations[,c(model,'AApos', 'Genename')]
  tmp$Model=model
  names(tmp) = c('Score','AApos','Gene',"Model")
  p = rbind(p,tmp)
}

p %>%
  filter(Gene == 'BRCA1') %>%
  ggplot(aes(x = AApos, y = Score)) +
  geom_point(size = 0.05) +
  geom_smooth(size = 0.3) +
  facet_wrap( ~Model, scales = "free")

ggsave(filename = "results/BRCA1.png", width = 16, height = 16, units = "in", dpi = 300)

p %>%
  filter(Gene == 'BRCA2') %>%
  ggplot(aes(x = AApos, y = Score)) +
  geom_point(size = 0.05) +
  geom_smooth(size = 0.3) +
  facet_wrap( ~Model, scales = "free")

ggsave(filename = "results/BRCA2.png", width = 16, height = 16, units = "in", dpi = 300)

rm(model, p, tmp)
```

Set some useful functions
```{r useful-functions}

# Create a filter to remove anything not in the BRCT/RING domains of BRCA1 and the DNA binding domain of BRCA2
filter_to_target_regions = function(d, variant_search = FALSE, l = links)
{
  RING_prot = c(1, 109)
  BRCT_prot = c(1642, 1855)
  DNABD_prot = c(2479, 3192)
  
  if(variant_search == TRUE){
    d$Variant = as.character(d$Variant)
    d = merge(d, l)
  }
  
  #Filter BRCA1 variants
  d$X.CHROM = as.numeric(d$X.CHROM)
  d$AApos   = as.numeric(d$AApos)
  BRCA1_keep_idx = which(
    d$X.CHROM ==  17 & (
      dplyr::between(d$AApos, RING_prot[1], RING_prot[2]) | dplyr::between(d$AApos,BRCT_prot[1], BRCT_prot[2]) 
      )
  )
  
  BRCA2_keep_idx =  which(d$X.CHROM ==  13 & dplyr::between(d$AApos, DNABD_prot[1], DNABD_prot[2]))
  d = d[c(BRCA1_keep_idx,BRCA2_keep_idx),]
  d = distinct(d, Variant, .keep_all = TRUE)
  rm(RING_prot, BRCT_prot, DNABD_prot, BRCA1_keep_idx, BRCA2_keep_idx)
  return(d)
}


### Define the MCC function
mcc <- function(predicted, actual)
{
  TP <- sum(actual ==   "Damaging" & predicted ==   "Damaging")
  TN <- sum(actual ==   "Neutral"     & predicted ==   "Neutral")
  FP <- sum(actual ==   "Neutral"     & predicted ==   "Damaging")
  FN <- sum(actual ==   "Damaging" & predicted ==   "Neutral")
  mcc <- ((TP*TN) - (FP*FN)) / sqrt((as.numeric(TP + FP)*as.numeric(TP + FN)*as.numeric(TN + FP)*as.numeric(TN + FN)))
  rm(TP, TN, FP, FN)
  return(round(mcc, digits = 3))
}

  

links = all_annotations %>% 
  mutate(Variant = paste(Chr,Pos, Ref, Alt, sep = "_")) %>%
  mutate(X.CHROM=Chr, POS=Pos, REF=Ref, ALT=Alt) %>%
  select(X.CHROM, AApos, Variant )
```



Load all functional data
```{r get-functional-data}
#Read in our data
neutral_list = read.csv("sources/neutral.list",sep = "\t",header = FALSE, stringsAsFactors = FALSE)
damaging_list = read.csv("sources/deleterious.list",sep = "\t",header = FALSE, stringsAsFactors = FALSE)

Hart_df = rbind(neutral_list,damaging_list)
names(Hart_df) = c("Gene","Variant","Call", "Source")

Hart_df = filter_to_target_regions(Hart_df, variant_search = TRUE)

rm(neutral_list, damaging_list)


# Read in annotations from CAVA
CAVA_BRCA1 = read.csv('sources/CAVA_BRCA1.txt',header = FALSE,
                   sep = "\t", stringsAsFactors = FALSE, 
                   na.strings = c('.'))
names(CAVA_BRCA1)=c('CSN','Gene','cpra','variantID', 'refAA','AApos', 'altAA')

# Read in annotations from CAVA
CAVA_BRCA2 = read.csv('sources/CAVA_BRCA2.txt',header = FALSE,
                   sep = "\t", stringsAsFactors = FALSE, 
                   na.strings = c('.'))
names(CAVA_BRCA2)=c('CSN','Gene','cpra','variantID', 'refAA','AApos', 'altAA')

```

```{r starita}
#######################################################################################################################################
#Read in Starita
#Get data
HDR_damaging_max = 0.33
HDR_neutral_min = 0.77

# Update 05/26/2019: Start from Startia et al Table S7, as an input
Starita = read.csv('sources/StaritaST7.tsv.txt',header = TRUE,
                   sep = "\t", stringsAsFactors = FALSE, 
                   na.strings = c('.'))

# Need to merge annotations so I can cross-reference
Starita = merge(Starita, CAVA_BRCA1, by='variantID')
# Remove duplicate variants
Starita = Starita %>%
  select(Starita_HDR_predict, variantID, cpra) %>%
  unique()
Starita$HDR=as.numeric(Starita$Starita_HDR_predict)

Starita = Starita %>%
  filter(HDR <=   HDR_damaging_max | HDR >=  HDR_neutral_min)



Starita_df = data.frame(
  Variant = as.character(Starita$cpra),
  Gene = 'BRCA1',
  Call = ifelse(Starita$HDR <=  HDR_damaging_max, "Damaging", 'Neutral' )
)
Starita_df = na.omit(Starita_df)

Starita_df = Starita_df %>%
  select(Gene, Variant, Call) %>%
  mutate(Source = "Starita et al., 2015") %>%
  unique()

Starita_df    = filter_to_target_regions(Starita_df, variant_search = TRUE)

rm(Starita, HDR_damaging_max, HDR_neutral_min)
```


```{r fernandez}
#######################################################################################################################################
#Read in Fernandes
#Get data
Fernandes = read.csv('sources/Fenandez.tsv', header = TRUE, 
                     sep = "\t", stringsAsFactors = FALSE)
#Remove WT and splic variants
omit = Fernandes$variant[grep('del|\\+|WT|\\/|ins',Fernandes$variant)]

Fernandes = Fernandes %>%
  filter(!variant %in% omit) %>%
  filter(IARC %in% c(0,1,4,5)) %>%
  mutate(Call = ifelse(IARC > 2, "Damaging", "Neutral"), Gene = 'BRCA1') %>%
  select(Gene, Variant, Call) %>%
  mutate(Source = "Fernandes et al., 2019") %>%
  unique()

Fernandes_df    = filter_to_target_regions(Fernandes, variant_search = TRUE)

rm(Fernandes, omit)

```

```{r findlay}
#######################################################################################################################################
#Read in Findlay
#Get data

# Update 05/26/2019: Start from Findlay et al Table S1, as an input

Findaly = read.csv('sources/Findlay_ST1.txt', header = TRUE,sep = "\t", 
                   stringsAsFactors = FALSE, na.strings = c('','NA','.'))
Findaly = Findaly %>% 
  mutate(cpra=paste(chromosome,`position..hg19.`,reference,alt,sep="_" ))
# Merge with annotations
Findaly = merge(Findaly, CAVA_BRCA1, by='cpra')


# Exclude splicing (1/28)
Findaly = Findaly %>%
  filter(consequence=='Missense', func.class %in% c('FUNC','LOF')) 

AlreadySeen = c(Fernandes_df$Variant, Starita_df$Variant)
Findaly$AlreadySeen = as.numeric(Findaly$cpra %in% AlreadySeen)

Fin_df = data.frame(
  Gene = Findaly$Gene,
  Variant = Findaly$cpra,
  Call = Findaly$func.class,
  Score = Findaly$function.score.mean,
  Seen = Findaly$AlreadySeen)

Findlay_df = Fin_df %>%
  filter(Seen ==  0)

Findlay_df$Call =  ifelse(Findlay_df$Call ==  'FUNC', 'Neutral', 'Damaging')

Findlay_df = Findlay_df %>%
  select(Gene, Variant, Call) %>%
  mutate(Source = "Findlay et al., 2018") %>%
  unique()

Findlay_df = filter_to_target_regions(Findlay_df, variant_search = TRUE)

rm(Fin_df, Findaly)
```

```{r combine-inputs, results='as-is'}
#######################################################################################################################################
# Combine experimental datasets
experimental_data = rbind(Hart_df, Fernandes_df, Starita_df, Findlay_df)
d = as.data.frame(table(experimental_data$Gene, experimental_data$Call, experimental_data$Source))
names(d) = c('Gene', 'Call', 'Source', 'RawCount')

experimental_data = distinct(experimental_data, Gene, Variant, .keep_all = TRUE)
e = as.data.frame(table(experimental_data$Gene, experimental_data$Call, experimental_data$Source))
names(e) = c('Gene', 'Call', 'Source', 'FilteredCount')

Counts = cbind(d, e$FilteredCount)
names(Counts)[length(names(Counts))] = 'FilteredCount'


rm(Hart_df, Fernandes_df, Starita_df, Findlay_df)
table(experimental_data$Source, experimental_data$Call,experimental_data$Gene)
table(experimental_data$Call,experimental_data$Gene)
```

More custom functions

````{r custom-functions}
prepare_inputs = function(data = experimental_data, 
                          annotations = all_annotations, 
                          gene = 'BRCA1', 
                          sources = NULL){

  s = NULL

  if ('Hart' %in% sources){s = c("This paper", "Guidugli et al., 2018", 
                                 "Hart et al, 2018", "Lee et al., 2010",
                                 "Lindor et al., 2012", "Woods et al., 2016", 
                                  "Guidugli et al., 2014")}
  if ('Fernandes' %in% sources){s = c(s, 'Fernandes et al., 2019')}
  if ('Starita' %in% sources){s = c(s, 'Starita et al., 2015')}
  if ('Findlay' %in% sources){s = c(s, 'Findlay et al., 2018')}
  
  # if sources = PATH, then take all pathogenic variants from all studies and benign from our study
  if(is.null(s)){
    data = data %>%
      filter(Gene == gene)
  }else{
    # Filter to known results for a specific gene
    data = data %>%
      filter(Source %in% s, Gene == gene)
  }
  
  data = distinct(data, Gene, Variant, .keep_all = TRUE)
  
  #Merge to get annotations
  data = merge(data,
    annotations,
    all.x = TRUE,
    by = 'Variant')
  
  # Sort list randomly
  data = sample(data)
  if ('Gene.x' %in% names(data)){ data = data %>% mutate(Gene = Gene.x, Gene.x=NULL, Gene.y=NULL)}
  if ('AApos.x' %in% names(data)){ data = data %>% mutate(AApos = AApos.x, AApos.x=NULL, AApos.y=NULL)}
  if ('X.CHROM.x' %in% names(data)){ data = data %>% mutate(X.CHROM = X.CHROM.x, X.CHROM.x=NULL, X.CHROM.y=NULL)}
  return(data)
}

split_data = function(known_scores, validation_fraction = 0.2 ){
  set.seed(999)
  known_scores$Call = as.factor(known_scores$Call)
  num_predictions = nrow(known_scores)

  #Create training testing and validation
  Neutral_idx = which(known_scores$Call == 'Neutral')
  Damaging_idx = which(known_scores$Call == 'Damaging')
  
  n_train_idx = sample(Neutral_idx, size = length(Neutral_idx) * 0.8)
  d_train_idx = sample(Damaging_idx, size = length(Damaging_idx) * 0.8)

  # Subset to training and validation
  idx = c(n_train_idx, d_train_idx)
  #Shuffle randomly
  idx = sample(idx)
  
  train_t = known_scores[idx,]
  test_t = known_scores[-idx,]
  splits = NULL
  splits$test = NULL
  splits$train = NULL
  splits$train = train_t
  splits$test = test_t
  return(splits)
}



k_fold_split_data = function(known_scores, k=10 ){
  set.seed(999)
  known_scores$Call = as.factor(known_scores$Call)
  num_predictions = nrow(known_scores)
  
  #Create training testing and validation
  Neutral_idx = which(known_scores$Call == 'Neutral')
  Neutral_fold = createFolds(Neutral_idx, k = k, list = FALSE)
  Damaging_idx = which(known_scores$Call == 'Damaging')
  Damaging_fold = createFolds(Damaging_idx, k = k, list = FALSE)
  
  
  # Subset to training and validation
  idx_f = c(Neutral_fold, Damaging_fold)
  
  known_scores$Fold = idx_f
  
  splits = sample(known_scores)
  return(splits)
}


train_model = function(splits, k_=NULL, Gene = 'BRCA1'){
  knitr::opts_chunk$set(echo=FALSE)

  h2o.init(min_mem_size = "8g", max_mem_size = "12g", nthreads = 10)
  
  train_ = as.h2o(splits$train)
  
  all_h2o_models =  h2o.automl(x = COLUMNS, y = "Call",
                  training_frame = train_,
                  keep_cross_validation_predictions = TRUE,
                  keep_cross_validation_models=TRUE,
                  keep_cross_validation_fold_assignment=TRUE,
                  nfolds = 5,
                  max_runtime_secs = training_time_limit,
                  sort_metric = 'mean_per_class_error',
                  stopping_metric = 'mean_per_class_error',
                  seed = 100)
  knitr::opts_chunk$set(echo=TRUE)
  
  #Save the model
  best_model = h2o.saveModel(object = all_h2o_models@leader,path = "results/",force = TRUE)
  best_model
  cv_predictions = h2o.getFrame(all_h2o_models@leader@model[["cross_validation_holdout_predictions_frame_id"]][["name"]])
  cv_groups = h2o.getFrame(all_h2o_models@leader@model[["cross_validation_fold_assignment_frame_id"]][["name"]])
  # Load model
  model_out = h2o.loadModel(best_model)
  model_vi  = as.data.frame(all_h2o_models@leader@model$variable_importances)
  threshold = h2o.find_threshold_by_max_metric(h2o.performance(model_out), 'absolute_mcc')
  my_list   = list('all_h2o_models' = all_h2o_models, 'best_model' = best_model, 
                   'model' = model_out, 'vi' =  model_vi, 'threshold' = threshold, 
                   'cv_predictions' = as.data.frame(cv_predictions), cv_groups = as.data.frame(cv_groups))
  return(my_list)
  }


test_model = function(splits, model, threshold = NULL){

  knitr::opts_chunk$set(echo=FALSE)
  h2o.init(min_mem_size = "16g", max_mem_size = "32g", nthreads = 1)

  test = as.h2o(splits$test)
  
  pred = h2o.predict(model, 
                     newdata = test, 
                     threshold = threshold)
  knitr::opts_chunk$set(echo=TRUE)

  pred = as.data.frame(pred)

  #Edit prediction based on previous threshold
  final_result = cbind(splits$test, pred)

  cm = confusionMatrix(final_result$predict,final_result$Call, positive = 'Damaging')
  m = mcc(final_result$predict,final_result$Call)
  my_list = list('result' = final_result, 'cm' = cm, 'MCC' = m)
  return(my_list)
}

apply_model_to_gene = function(model = model, threshold = threshold,
                               gene = 'BRCA1', annotations = all_annotations){
  #Run model on each gene
  tmp = annotations %>%
    filter(Genename == gene)
  pred = h2o.predict(model, newdata = as.h2o(tmp), threshold = threshold)
  pred = as.data.frame(pred)
  tmp$Prediction = pred$predict
  tmp$`BRCA-ML` = pred$Damaging
  tmp$Neut = pred$Neutral
  return(tmp)
}


plot_full_gene = function(data, title, splits = splits, model_name='BRCA-ML', t_=NULL){
  
  #Pull out training, validation, & testing variants to plot
  
  if(length(splits$train)){
    variants = splits$train %>%
      select(AApos, Call)
    
  
    variants = rbind(variants,splits$test %>%
      select(AApos, Call)
    )  
  }else{
    variants = NULL
  }
  
  tmp = data[,c("Genename","BRCA-ML","AApos","Prediction")]
  tmp$Model = model_name
  names(tmp) = c("Gene","Score","AApos","Prediction","Model")
  tmp$Prediction = as.character(tmp$Prediction)
  if (!is.null(t_)){tmp = tmp %>% mutate(Prediction = ifelse(Score > t_, 'Damaging', 'Neutral'))}
  p = tmp %>%
    ggplot(aes(x = AApos,y = Score, colour = Prediction)) +
    geom_point(size = 0.2) +
    #facet_grid(Model ~ Gene, scales = 'free') +
    geom_smooth(aes(x = AApos,y = Score), inherit.aes = FALSE) +
    ggtitle(title) +
    theme(legend.position="none") 
  if(length(splits$train)){
    p = p + geom_rug(
        aes(x=AApos, colour=Call, alpha = 1/2), 
        inherit.aes = FALSE, 
        sides = "b", 
        data = variants)
  }
  return(p)
}

```

```{r}
  get_cv_estimate = function(gene='BRCA1'){
    h2o.init(min_mem_size = "8g", max_mem_size = "10g", nthreads = 36)
    # Nested cross validation of the "pick the leader" process
    #bx_input = b1_DATA_SPLITS$train
    if (gene == 'BRCA1'){
      bx_input = b1_DATA_SPLITS$train
    }else{
      bx_input = b2_DATA_SPLITS$train
    }
    bx_DATA_SPLITS = k_fold_split_data(bx_input, k = v_fold_nest)  # change to k-fold CV, 2020-05-28
    
    preds = NULL
    for(v in 1:v_fold_nest) {
        train_v <- bx_DATA_SPLITS[which(bx_DATA_SPLITS$Fold != v), ]
        test_v <- bx_DATA_SPLITS[which(bx_DATA_SPLITS$Fold == v), ]
        
        train = as.h2o(train_v)
        all_h2o_models = h2o.automl(x = COLUMNS, y = "Call",
                        training_frame = train,
                        keep_cross_validation_predictions = FALSE, # don't need these
                        nfolds = 5,  # should be increased to 10?
                        max_runtime_secs = training_time_limit,
                        algo_parameters = list(list(min_rows=-1)),
                        sort_metric = 'mean_per_class_error',
                        stopping_metric = 'mean_per_class_error',
                        seed = 100)
        test = as.h2o(test_v)
        pred = h2o.predict(all_h2o_models@leader, newdata = test)
        pred= as.data.frame(pred)
        pred = cbind(pred, test_v)
        preds = rbind(preds, pred)

        write.csv(preds, file=paste0('results/', gene,'_res.fold',v,'.csv'))
    }
    return(preds)
  }


```

Define columns to TRAIN on

```{r get-thresholds}
models = read.csv('sources/base_thresholds', header = FALSE, sep="\t", stringsAsFactors = FALSE)
COLUMNS = models$V1
```

Prepare inputs for the training
```{r data_prep}
#BRCA1
b1_input = prepare_inputs(gene = 'BRCA1')
b1_DATA_SPLITS = split_data(b1_input)

b2_input = prepare_inputs(gene = 'BRCA2')
b2_DATA_SPLITS = split_data(b2_input)
```

Calculate nested cross validation metrics
```{r b1-cv}
#Add in metrics from our current model
if (train==TRUE){
  b1_cv = get_cv_estimate(gene = 'BRCA1') # WARNING! THIS TAKES AN EXTREMELY LONG TIME!!
  b2_cv = get_cv_estimate(gene = 'BRCA2') # WARNING! THIS TAKES AN EXTREMELY LONG TIME!!
} else{
  for (v in 1:v_fold_nest){
    b1_cv = NULL
    b2_cv = NULL
    b1 = read.csv(paste0('results/','BRCA1_res.fold',v,'.csv'))
    b1_cv = rbind(b1_cv, b1)
    b2 = read.csv(paste0('results/','BRCA2_res.fold',v,'.csv'))
    b2_cv = rbind(b2_cv, b2)
    b1_cv = na.omit(b1_cv)
    b2_cv = na.omit(b2_cv)
  }
}
```

Train
```{r train_brca1, echo = FALSE}
if (train==TRUE){
  b1_TRAINED_DATA = train_model(b1_DATA_SPLITS)
}else{
  h2o.init()
  model_out = h2o.loadModel('results/GBM_grid__1_AutoML_20200622_080955_model_22')
  model_vi  = as.data.frame(model_out@model$variable_importances)
  threshold = h2o.find_threshold_by_max_metric(h2o.performance(model_out), 'absolute_mcc')
  
  b1_TRAINED_DATA   = list('all_h2o_models' = NULL, 'best_model' = model_out@model_id, 
                   'model' = model_out, 'vi' =  model_vi, 'threshold' = threshold, 
                   'cv_predictions' = b1_cv)

}
b1_TESTED_DATA = test_model(b1_DATA_SPLITS, b1_TRAINED_DATA$model, threshold = b1_TRAINED_DATA$threshold)
b1_PREDICTIONS = apply_model_to_gene(model = b1_TRAINED_DATA$model, 
                                          gene = 'BRCA1', 
                                          threshold = b1_TRAINED_DATA$threshold)
b1_PLOTS = plot_full_gene(b1_PREDICTIONS, 
                               'BRCA1', 
                               splits = b1_DATA_SPLITS, 
                               t_ = b1_TRAINED_DATA$threshold )
```

```{r train_brca2, echo = FALSE}

#BRCA2

if (train==TRUE){
  b2_TRAINED_DATA = train_model(b2_DATA_SPLITS)
  }else{
   h2o.init()
  model_out = h2o.loadModel('results/GBM_grid__1_AutoML_20200622_085730_model_23')
  model_vi  = as.data.frame(model_out@model$variable_importances)
  threshold = h2o.find_threshold_by_max_metric(h2o.performance(model_out), 'absolute_mcc')
  # Get CV predictions
  cv_predictions=NULL
  cv_groups=NULL
  b2_cv_holdouts=NULL
  for (i in 1:v_fold_nest){
    tmp = read.csv(paste0('results/BRCA2_res.fold',i,'.csv'))

    test = as.h2o(tmp)
    pred = h2o.predict(model_out, newdata = test)
    tmp = tmp %>% select('Call','Variant')
    tmp_df = data.frame(Call=tmp$Call, Variant=tmp$Variant, cv_pred=as.data.frame(pred)[, "Damaging"],Fold=i)
    cv_predictions = rbind(cv_predictions, tmp_df)
    b2_cv_holdouts = rbind(b2_cv_holdouts, as.data.frame(pred))
    cv_groups = rbind(cv_groups, data.frame(fold_assignment=rep(i,nrow(tmp))))
  }
  b2_TRAINED_DATA   = list('all_h2o_models' = NULL, 'best_model' = model_out@model_id, 
                   'model' = model_out, 'vi' =  model_vi, 'threshold' = threshold, 
                   'cv_predictions' = as.data.frame(cv_predictions), cv_groups = as.data.frame(cv_groups))
  b2_cv_holdouts$cv_pred = b2_cv_holdouts$Damaging
  }


b2_TESTED_DATA = test_model(b2_DATA_SPLITS, b2_TRAINED_DATA$model, threshold = b2_TRAINED_DATA$threshold)
b2_PREDICTIONS = apply_model_to_gene(model = b2_TRAINED_DATA$model, 
                                          gene = 'BRCA2', 
                                          threshold = b2_TRAINED_DATA$threshold)
b2_PLOTS = plot_full_gene(b2_PREDICTIONS, 
                               'BRCA2', 
                               splits = b2_DATA_SPLITS, 
                               t_ = b2_TRAINED_DATA$threshold )

```

Make functions to determine cutoff for other missense predictors

```{r calculate-cutoffs}
get_res = function(input_data, model){
  
  ## Use the training data to get an optimal threshold
  
  # Make a df that merges the experimental results with predictors
  tmp = input_data$train[,c('Call',model)]
  roc_obj <- roc(tmp$Call, tmp[,model])
  d = as.data.frame(coords(roc_obj, "best", ret=c("threshold","sensitivity", "specificity", "tp", "tn", "fp", "fn")))
  threshold = unlist(d[1])
  if(is.null(threshold)){threshold=0.5}
  ################################################################################
  ## 05/28/20
  tmp$Model='Neutral'
  if (model %in% c("SiftScore")){
    tmp$Model = ifelse(tmp[,model]>threshold,'Neutral','Damaging')
  }else{
    tmp$Model = ifelse(tmp[,model]<threshold,'Neutral','Damaging')  
  }
  cm = confusionMatrix(as.factor(tmp$Model), as.factor(tmp$Call), positive="Damaging")
  tp=cm$table[1]
  fn=cm$table[2]
  fp=cm$table[3]
  tn=cm$table[4]
  sensitivity=cm$byClass[1]
  specificity=cm$byClass[2]
  # Add no call information
  missingness = (nrow(tmp)-sum(cm$table))/nrow(tmp) * 100
  d = data.frame(tp=tp,fp=fp,fn=fn,tn=tn,sensitivity=sensitivity,specificity=specificity, threshold=threshold, model=model, missingness=missingness)
  ################################################################################
  rownames(d)=NULL
  d = as.data.frame(d)
  d$mcc = ((d$tp*d$tn) - (d$fp * d$fn)) / sqrt((as.numeric(d$tp + d$fp)*as.numeric(d$tp + d$fn)*as.numeric(d$tn + d$fp)*as.numeric(d$tn + d$fn)))
  d$model = model
  pr_1 =NULL
  # Add precision recall
  
  idx = which(tmp$Call=='Neutral')
  tryCatch(
    {
      pr_1 = pr.curve(
              scores.class0 = tmp[-idx, model],
              scores.class1 = tmp[idx, model], 
              curve=TRUE, sorted = FALSE)
      
      
    },
    error = function(cond){
      pr_1 = NULL
    })
    
    if (is.null(pr_1)){
      proc = NULL
      proc$V1 = 0
      proc$V2 = 0
      proc$V3 = 0
      proc$Model = model
      d$PRauc = 0
    }else{
      d$PRauc = pr_1$auc.integral
      proc = pr_1$curve %>% as.data.frame()
      proc$Model = model
    }

  my_list = list('roc' = roc_obj, 'res' = d, 'proc' = proc)
  rm(d, tmp)
  return(my_list)
  
}

compute_optimal = function(input_data=b1_DATA_SPLITS){
  RES = NULL
  ROC = NULL
  PROC = NULL
  for (model in models$V1){
    l = get_res(input_data, model)
    RES = rbind(RES, l$res)
    d1 = data.frame(sensitivities=l$roc$sensitivities, specificities=l$roc$specificities, auc = as.numeric(l$roc$auc))
    d1$Model = model
    ROC = rbind(ROC, d1)
    
    # Add PROC
    tryCatch(
      {PROC = rbind(PROC, l$proc)},
      error = function(cond){
        PROC = PROC
      }
    )
  }
  my_list = list('RES' = RES, 'ROC' = ROC, 'PROC' = PROC)
  return(my_list)
}
```


Compute optimal for BRCA1
```{r compute-brca1-optimal}
res_roc_1 = compute_optimal(input_data=b1_DATA_SPLITS)
```

```{r merge-brca1}

# Get ROC/PROC curve for BRCA-ML on CV set
roc_obj <- roc(b1_cv$Call,b1_cv$Damaging)

# Get metrics from cv-set
cm = confusionMatrix(as.factor(b1_cv$Call), b1_cv$predict, positive="Damaging")
tp=cm$table[1]
fn=cm$table[2]
fp=cm$table[3]
tn=cm$table[4]
sensitivity=cm$byClass[1]
specificity=cm$byClass[2]
# Add no call information
missingness = (nrow(b1_DATA_SPLITS$train)-sum(cm$table))/nrow(b1_DATA_SPLITS$train) * 100
threshold = b1_TRAINED_DATA$threshold
model='BRCA-ML'
d = data.frame(tp=tp,fp=fp,fn=fn,tn=tn,sensitivity=sensitivity,specificity=specificity, threshold=threshold, model=model, missingness=missingness)
rownames(d)=NULL
d = as.data.frame(d)
d$mcc = ((d$tp*d$tn) - (d$fp * d$fn)) / sqrt((as.numeric(d$tp + d$fp)*as.numeric(d$tp + d$fn)*as.numeric(d$tn + d$fp)*as.numeric(d$tn + d$fn)))

# Add precision recall
idx = which(b1_cv$Call=='Neutral')
pr_1 = pr.curve(
  scores.class0 = b1_cv$Damaging[-idx],
  scores.class1 = b1_cv$Damaging[idx], 
  curve=TRUE,sorted = FALSE) 
d$PRauc = pr_1$auc.integral

res_roc_1$RES = rbind(res_roc_1$RES, d)

## Start getting curve info
d1 = data.frame(sensitivities=roc_obj$sensitivities, specificities=roc_obj$specificities, auc = as.numeric(roc_obj$auc))
d1$Model='BRCA-ML'

d2 = pr_1$curve %>% as.data.frame()
d2$Model = 'BRCA-ML'

res_roc_1$ROC = rbind(res_roc_1$ROC, d1)
res_roc_1$PROC = rbind(res_roc_1$PROC, d2)
rm(d,d1,d2, tmp)

p1 = res_roc_1$ROC %>%
  ggplot(aes(y=sensitivities, x=1-specificities, color = Model)) +
  geom_line() +
  scale_fill_brewer(palette="Paired") +
  labs(color='Model') +
  xlab("Specificity") +
  ylab("Sensitivity") +
  ggtitle('BRCA1') +
  theme(legend.position = "none")

```

Compute for BRCA2
```{r compute-brca2}
res_roc_2 = compute_optimal(input_data=b2_DATA_SPLITS)
```


```{r merge-brca2}
# Get ROC/PROC curve for BRCA-ML on CV set
roc_obj <- roc(b2_cv$Call, b2_cv$Damaging)

# Get metrics from test set 
cm = confusionMatrix(as.factor(b2_cv$Call), b2_cv$predict, positive="Damaging")
tp=cm$table[1]
fn=cm$table[2]
fp=cm$table[3]
tn=cm$table[4]
sensitivity=cm$byClass[1]
specificity=cm$byClass[2]
# Add no call information
missingness = (nrow(b2_TESTED_DATA$result)-sum(cm$table))/nrow(b2_TESTED_DATA$result) * 100
threshold = b2_TRAINED_DATA$threshold
model='BRCA-ML'
d = data.frame(tp=tp,fp=fp,fn=fn,tn=tn,sensitivity=sensitivity,specificity=specificity, threshold=threshold, model=model, missingness=missingness)
rownames(d)=NULL
d = as.data.frame(d)
d$mcc = ((d$tp*d$tn) - (d$fp * d$fn)) / sqrt((as.numeric(d$tp + d$fp)*as.numeric(d$tp + d$fn)*as.numeric(d$tn + d$fp)*as.numeric(d$tn + d$fn)))

# Add precision recall
idx = which(b2_cv$Call=='Neutral')
pr_2 = pr.curve(
  scores.class0 = b2_cv$Damaging[-idx],
  scores.class1 = b2_cv$Damaging[idx], 
  curve=TRUE,sorted = FALSE) 
d$PRauc = pr_2$auc.integral

res_roc_2$RES = rbind(res_roc_2$RES, d)

d1 = data.frame(sensitivities=roc_obj$sensitivities, specificities=roc_obj$specificities, auc = as.numeric(roc_obj$auc))
d1$Model='BRCA-ML'

d2 = pr_1$curve %>% as.data.frame()
d2$Model='BRCA-ML'


res_roc_2$ROC = rbind(res_roc_2$ROC, d1)
res_roc_2$PROC = rbind(res_roc_2$PROC, d2)
rm(d,d1,tmp)

p2 = res_roc_2$ROC %>%
  ggplot(aes(y=sensitivities, x=1-specificities, color = Model)) +
  geom_line() +
  scale_fill_brewer(palette="Paired") +
  labs(color='Model') +
  xlab("Specificity") +
  ylab("Sensitivity") +
  ggtitle('BRCA2') 

```


Make a combined image
```{r combine-p1-p2}
ggarrange(p1, p2,ncol=2, nrow=1, common.legend = TRUE, legend = "bottom")
```

Make PR Curves
```{r make-pr-curves}
res_roc_1$PROC$Gene='BRCA1'
res_roc_2$PROC$Gene='BRCA2'
PROC = rbind(res_roc_1$PROC, res_roc_2$PROC)

p3 = PROC %>% 
  filter(Gene == 'BRCA1') %>%
  ggplot(aes(x=V1, y=V2, color=Model)) + 
  geom_line() +
  facet_grid(. ~ Gene) +
  theme(legend.position = "none") +
  ylab('Precision') +
  xlab("Recall")

p4 = PROC %>% 
  filter(Gene == 'BRCA2') %>%
  ggplot(aes(x=V1, y=V2, color=Model)) + 
  geom_line() +
  facet_grid(. ~ Gene) +
  theme(legend.position = "right")+
  ylab('Precision') +
  xlab("Recall")
```

All together now.  This represent the performance on the cross validation data for BRCA-ML (overfit on  all others).
```{r plot-pr}
figure = ggarrange(p1, p2, p3, p4, common.legend = TRUE, legend = "right" )
annotate_figure(figure, fig.lab ="Cross validation performance",fig.lab.size=18, fig.lab.face = 'bold', top = text_grob("", size = 14))
ggsave(filename = "results/Combined_PROC_ROC.png", width = 14, height = 8, units = "in", dpi = 300)

```

Make pair plot
```{r pair-plot}
tmp = b1_PREDICTIONS %>%
  select(AApos, `BRCA-ML`, BayesDel) %>%
  mutate(Gene='BRCA1')

tmp2 = b2_PREDICTIONS %>%
  select(AApos, `BRCA-ML`, BayesDel) %>%
  mutate(Gene='BRCA2')
 
tmp = rbind(tmp, tmp2)
tmp = tmp %>%
  gather(key = "Model", value = 'Value',  `BRCA-ML`:BayesDel)


## Assign damaging score
tmp$Prediction='Neutral'
idx1 = which(tmp$Gene == 'BRCA1' & tmp$Value >= b1_TRAINED_DATA$threshold & tmp$Model == 'BRCA-ML')
idx2 = which(tmp$Gene == 'BRCA2' & tmp$Value >= b2_TRAINED_DATA$threshold & tmp$Model == 'BRCA-ML')

bd_thresh = res_roc_1$RES %>% filter(model=='BayesDel') %>% select(threshold) %>% as.numeric()
idx3 = which(tmp$Gene == 'BRCA1' & tmp$Value >= bd_thresh & tmp$Model == 'BayesDel')
bd_thresh = res_roc_2$RES %>% filter(model=='BayesDel') %>% select(threshold) %>% as.numeric()
idx4 = which(tmp$Gene == 'BRCA2' & tmp$Value >= bd_thresh & tmp$Model == 'BayesDel')

tmp$Prediction[c(idx1,idx2,idx3,idx4)] = 'Damaging'
rm(idx1,idx2,idx3,idx4, bd_thresh)

tmp %>%
  ggplot(aes(x=AApos, y=Value, color=Prediction)) +
  geom_point(size=0.5) +
  geom_smooth(aes(x = AApos,y = Value), inherit.aes = FALSE) +
  facet_wrap(~Model+Gene, scales = "free", ncol=2) +
  theme(legend.position = "bottom")

ggsave(filename = "results/PairPlot.png", width = 14, height = 8, units = "in", dpi = 300)
```

```{r}
get_metrics = function(truth_vector, predicted_class, model, num_rows, threshold=NA){
  cm = confusionMatrix(truth_vector, predicted_class, positive="Damaging")
  tp=cm$table[1]
  fn=cm$table[2]
  fp=cm$table[3]
  tn=cm$table[4]
  sensitivity=cm$byClass[1]
  specificity=cm$byClass[2]
  missingness = (num_rows-sum(cm$table))/num_rows * 100

  d = data.frame(tp=tp,fp=fp,fn=fn,tn=tn,missingness=missingness,sensitivity=sensitivity,specificity=specificity, threshold=threshold, model=model)
  rownames(d)=NULL
  d = as.data.frame(d)
  d$mcc = ((d$tp*d$tn) - (d$fp * d$fn)) / sqrt((as.numeric(d$tp + d$fp)*as.numeric(d$tp + d$fn)*as.numeric(d$tn + d$fp)*as.numeric(d$tn + d$fn)))
  return(d)
}
```

Calculate metrics on test set
```{r print-summary, echo=FALSE}
res_roc_1$RES$Gene='BRCA1'
metrics_table = NULL

for (i in 1:nrow(res_roc_1$RES)){
  model = res_roc_1$RES[i,'model']
  threshold = res_roc_1$RES[i,'threshold']
  num_rows = nrow(b1_DATA_SPLITS$test)
  if (model=='SiftScore'){
    tmp = b1_DATA_SPLITS$test %>% select_('Call', paste(model))
    names(tmp) = c('Call','Prediction')
    tmp = tmp %>% 
      mutate(Classification=ifelse(Prediction < threshold,'Damaging','Neutral')) %>%
      mutate(Classification=as.factor(Classification))
    tmp = get_metrics(tmp$Call, tmp$Classification, model, num_rows, threshold = threshold)
    tmp$Gene = 'BRCA1'
    metrics_table = rbind(metrics_table, tmp)
  }else if(model=='BRCA-ML'){
    tmp = test_model(b1_DATA_SPLITS, b1_TRAINED_DATA$model, threshold = threshold)
    tmp = get_metrics(tmp$result$Call, tmp$result$predict, 'BRCA-ML', num_rows, threshold = threshold)
    tmp$Gene = 'BRCA1'
    metrics_table = rbind(metrics_table, tmp)
    
  }else{
    tmp = b1_DATA_SPLITS$test %>% select_('Call', paste(model))
    names(tmp) = c('Call','Prediction')
    tmp = tmp %>% 
      mutate(Classification=ifelse(Prediction > threshold,'Damaging','Neutral')) %>%
      mutate(Classification=as.factor(Classification))
    tmp = get_metrics(tmp$Call, tmp$Classification, model, num_rows, threshold = threshold)
    tmp$Gene = 'BRCA1'
    metrics_table = rbind(metrics_table, tmp)
  }
}
rm(i, tmp)
```

```{r}
res_roc_2$RES$Gene='BRCA2'

for (i in 1:nrow(res_roc_2$RES)){
  model = res_roc_2$RES[i,'model']
  threshold = res_roc_2$RES[i,'threshold']
  num_rows = nrow(b2_DATA_SPLITS$test)

  if (model=='SiftScore'){
    tmp = b2_DATA_SPLITS$test %>% select_('Call', paste(model))
    names(tmp) = c('Call','Prediction')
    tmp = tmp %>% 
      mutate(Classification=ifelse(Prediction < threshold,'Damaging','Neutral')) %>%
      mutate(Classification=as.factor(Classification))
    tmp = get_metrics(tmp$Call, tmp$Classification, model, num_rows, threshold = threshold)
    tmp$Gene = 'BRCA2'
    metrics_table = rbind(metrics_table, tmp)
  }else if(model=='BRCA-ML'){
    tmp = test_model(b2_DATA_SPLITS, b2_TRAINED_DATA$model, threshold = threshold)
    tmp = get_metrics(tmp$result$Call, tmp$result$predict, 'BRCA-ML', num_rows, threshold = threshold)
    tmp$Gene = 'BRCA2'
    metrics_table = rbind(metrics_table, tmp)
    
  }else{
    tmp = b2_DATA_SPLITS$test %>% select_('Call', paste(model))
    names(tmp) = c('Call','Prediction')
    tmp = tmp %>% 
      mutate(Classification=ifelse(Prediction > threshold,'Damaging','Neutral')) %>%
      mutate(Classification=as.factor(Classification))
    tmp$Classification=factor(tmp$Classification,levels=c('Damaging','Neutral'))
    tmp = get_metrics(tmp$Call, tmp$Classification, model, num_rows, threshold = threshold)
    tmp$Gene = 'BRCA2'
    metrics_table = rbind(metrics_table, tmp)
  }
}
rm(i, tmp)

```

Add confidence intervals
```{r cis}
add_ci = function(true1, false1, false2){
  correct = true1
  incorrect = sum(false1,false2)
  ci_ = tryCatch({
    res = prop.test(x = correct, n = incorrect, correct = FALSE)
    lci = round(res$conf.int[1],3)
    uci = round(res$conf.int[2],3)
    ci_ = paste(lci, uci, sep="-")
    return(ci_)
  },
  error=function(cond){
    return("-")
  }
  )
}
metrics_table$sensitivity_ci = NULL
metrics_table$specificity_ci = NULL
for (r in 1:nrow(metrics_table)){
  metrics_table$sensitivity_ci[r] = add_ci(metrics_table$tp[r], metrics_table$tp[r], metrics_table$fn[r])
  metrics_table$specificity_ci[r] = add_ci(metrics_table$tn[r], metrics_table$tn[r], metrics_table$fp[r])
}

```

```{r  metrics, results='as-is'}
metrics_table %>% 
  filter(missingness < 1) %>%
  arrange(-mcc) %>%
  mutate(mcc=round(mcc,2)) %>%
  kable()
```


```{r}
model_list = as.character(metrics_table$model)[order(as.character(metrics_table$model))]
model_list = unique(model_list)
metrics_table$model = factor(metrics_table$model, levels=model_list)
metrics_table %>%  
  ggplot(aes(x=model, y=mcc, fill=Gene)) +
  geom_bar(stat='identity') +
  facet_grid(Gene~.) + 
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  ylab('MCC') +
  xlab('')

ggsave(filename = "results/MCC_plot.png", width = 14, height = 6, units = "in", dpi = 300)
```

Add HGVS Nomenclature
```{r}
NOMENCLATURE = read.csv('sources/Nomenclature',sep='\t')
TABLE_S2 = rbind(b1_PREDICTIONS, b2_PREDICTIONS)

TABLE_S2 = TABLE_S2 %>%
  merge(x=TABLE_S2, y=NOMENCLATURE, by.x='Variant', by.y='ID')
```


Save table S2.
```{r write-table}
# Update BRCA-ML Label to contain a version number
version='1.1'
idx1=which(names(TABLE_S2)=='BRCA-ML')
idx2=which(names(TABLE_S2)=='Prediction')
names(TABLE_S2)[idx1] = paste('BRCA-ML', version, sep='.')
names(TABLE_S2)[idx2] = paste('BRCA-ML', version, 'Prediction', sep='.')

write.table(TABLE_S2, file='results/TableS2.tsv', sep='\t', row.names = FALSE)
```

Show Variable Importances for BRCA1
```{r b1_vi}
b1_TRAINED_DATA$model@model$model_summary
kable(b1_TRAINED_DATA$vi)
```

Show Variable Importances for BRCA2
```{r}
b2_TRAINED_DATA$model@model$model_summary
kable(b2_TRAINED_DATA$vi)
```

```{r save}
#save.image(file="working/v3.4_rel2.RData")
sessionInfo()
```